src/services/index.ts

```diff
@@ -0,0 +1,448 @@
+    await voidTransaction(mysql, hash, addressBalanceMap);
+    await markUtxosAsVoided(mysql, dbTxOutputs);
+
+    const addresses = Object.keys(addressBalanceMap);
+    await validateAddressBalances(mysql, addresses);
+
+    await dbUpdateLastSyncedEvent(mysql, fullNodeEvent.event.id);
+
+    logger.debug(`Voided tx ${hash}`);
+
+    await mysql.commit();
+  } catch (e) {
+    logger.debug(e);
+    await mysql.rollback();
+
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const handleUnvoidedTx = async (context: Context) => {
+  const mysql = await getDbConnection();
+  await mysql.beginTransaction();
+
+  try {
+    const fullNodeEvent = context.event as FullNodeEvent;
+
+    const { hash } = fullNodeEvent.event.data;
+
+    logger.debug(`Tx ${hash} got unvoided, cleaning up the database.`);
+
+    await cleanupVoidedTx(mysql, hash);
+
+    logger.debug(`Unvoided tx ${hash}`);
+
+    await mysql.commit();
+  } catch (e) {
+    logger.debug(e);
+    await mysql.rollback();
+
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const handleTxFirstBlock = async (context: Context) => {
+  const mysql = await getDbConnection();
+  await mysql.beginTransaction();
+
+  try {
+    const fullNodeEvent = context.event as FullNodeEvent;
+
+    const {
+      hash,
+      metadata,
+      timestamp,
+      version,
+      weight,
+    } = fullNodeEvent.event.data;
+
+    let height: number | null = metadata.height;
+
+    if (!metadata.first_block) {
+      height = null;
+    }
+
+    await addOrUpdateTx(mysql, hash, height, timestamp, version, weight);
+    await dbUpdateLastSyncedEvent(mysql, fullNodeEvent.event.id);
+    logger.debug(`Confirmed tx ${hash}: ${fullNodeEvent.event.id}`);
+
+    await mysql.commit();
+  } catch (e) {
+    logger.error('E: ', e);
+    await mysql.rollback();
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const updateLastSyncedEvent = async (context: Context) => {
+  const mysql = await getDbConnection();
+
+  const lastDbSyncedEvent: LastSyncedEvent | null = await getLastSyncedEvent(mysql);
+  // @ts-ignore
+  const lastEventId = context.event.event.id;
+
+  if (lastDbSyncedEvent
+     && lastDbSyncedEvent.last_event_id > lastEventId) {
+     logger.error('Tried to store an event lower than the one on the database', {
+       lastEventId,
+       lastDbSyncedEvent: JSON.stringify(lastDbSyncedEvent),
+     });
+      mysql.destroy();
+     throw new Error('Event lower than stored one.');
+  }
+  await dbUpdateLastSyncedEvent(mysql, lastEventId);
+
+  mysql.destroy();
+};
+
+export const fetchInitialState = async () => {
+  const mysql = await getDbConnection();
+  const lastEvent = await getLastSyncedEvent(mysql);
+
+  mysql.destroy();
+
+  return { lastEventId: lastEvent?.last_event_id };
+};
+
+export const validateNetwork = async () => {
+  return;
+};
```

Comment:
Missing implementation?

Reply:
Refactored in 517953574c153b62a912b6505bcd6259ef5766e4